### ISSUE TITLE
Get field of record type using dot (.)

### ISSUE THREAD

**Issue by Gabriel439** [dhall-lang/dhall-lang#1079]
The idea for this comes from [this Stack Overflow question](https://stackoverflow.com/questions/64165340/dhall-repl-how-to-get-the-type-of-field-of-record-type)

The basic idea is that the `.` operator would now work on record types to access the types of their fields.  For example:

```dhall
let Person = { name : Text, age : Natural }

in  Person.name  -- Text
```

**Comment by mheiber** [comment]
This case is similar to type-level merge (`//\\`) and value-level recursive merge (`/\`). In that case Dhall has separate symbols for the two operations. Why would Dhall need separate symbols for type-vs-value merging, but only a single symbol (`.`) for field access? (genuine question, I don't know the answer)

**Comment by Gabriel439** [comment]
@mheiber: No good reason 🙂.  We could probably unify `∧` and `⩓` into a single operator

**Comment by mheiber** [comment]
Would projecting multiple fields work?

```dhall
let Person = { name : Text, age : Natural, height : Natural }
in  Person.{name, age}  -- { name : Text, age: Natural, height : Natural }
```

**Comment by Gabriel439** [comment]
@mheiber: Yeah, I assume that projection would work, too

**Comment by mheiber** [comment]
```
Γ, (z : e) ⊢ z.x : T
──────────────────────
e.x : T

Γ, (z : e) ⊢ z.{xs...} : T
──────────────────────
e.{xs...} : T
```

Here are alternate rules that don't use term-level dot. I struggled with these:

```
Γ ⊢ e : Type
e ⇥ { x : T, xs… }
Γ ⊢ T : t₀
──────────────────
e.x : t₀

Γ ⊢ e : Type
Γ ⊢ {} : t
───────────
e.{} : t

Γ ⊢ e : Type
e ⇥ { x : t₀, xs… }
Γ ⊢ e.{ xs… } : ts₁
{ x : t₀ } ⋁ ts₁ = t₂
──────────────────────────────────────────────────  ; x ∉ { xs… }
e.{ x, xs… } : t₂
```

I'm a bit confused about how judgments for type-level operators work. In the following rule, for example, we know that l is a type, but which type is it? How is the goal different from   `Γ ⊢ l ⩓ r : Type`?

```
    Γ ⊢ l : t
    Γ ⊢ r : Type
    l ⇥ { ls… }
    r ⇥ {}
    ─────────────
    Γ ⊢ l ⩓ r : t
```

**Comment by Gabriel439** [comment]
I think for the first rule you don't necessarily want to enforce that `e` is a `Type`.  For example, if the record stores a type then the type of the record type could be `Kind`:

```
Γ ⊢ e : c₀
e ⇥ { x : T, xs… }
Γ ⊢ T : c₁
──────────────────
e.x : c₁
```

In the type-checking rule for `l ⩓ {}` we don't necessarily know that `t` is `Type`.  For example, `{ x : Type } ⩓ {}` is valid Dhall code and in that scenario `t` is `Kind`.

For that rule, what I believe you want is:

```
Γ ⊢ e : Type
e ⇥ { ts… }
───────────
e.{} : Type
```

In other words, just check that `e` is a record type, and the result is always `Type`.  Then the second rule can take care of finding the least upper bound over all of the constants:

```
Γ ⊢ e : c
e ⇥ { x : T, xs… }
Γ ⊢ { x : T } : c₀  ; Maybe this rule can be simplified further
Γ ⊢ e.{ xs… } : c₁
c₀ ⋁ c₁ = c₂
───────────────────  ; x ∉ { xs… }
e.{ x, xs… } : c₂
```

**Comment by Grissess** [comment]
Hello! I see this has been open for a little while, and I've run into something similar to what prompted the SO question with `dhall-kubernetes`, though not as directly. In effect:

```dhall
let TheType = {
    -- ...
    foo : List SomeObscenelySpecificTypeExpr,
    bar : List AnotherDifferentSpecificTypeExpr,
    -- ....
}

let default = {
    -- ...
    foo = ([] : TheType.foo),
    bar = ([] : TheType.bar),
    -- ...
}

in { Type = TheType, default }
```

I could surely refactor those types out (at risk of having to find a good name for them), but it surprised me that this didn't work as intended.

Would a hand with any part of getting this working be appreciated?

**Comment by winitzki** [comment]
I have been looking again at the syntax enhancements for records / record types. I believe the syntax `x.{...}` is very useful but there is one corner case: `x.{}` which is problematic. I believe the best solution is to disallow that syntax entirely. I don't see any use for `x.{}`: we know in advance what the result should be. Also, `x.r` with `r = {}` is not valid since `x.r` is the syntax for record field access. So, I'm going to take up this work again and implement rules that completely disallow the syntax `x.{}` for record types `x`.

I would also like to disallow `x.{}` for record values `x` because it's also not useful: we know it's `{=}` for all record values `x`. But that would be technically a breaking change.
